### ISSUE TITLE
RFC: Extract Injector.js as an NPM module

### ISSUE THREAD

**Issue by Cheddam** [silverstripe/silverstripe-admin#836]
Injector.js is the foundation of the pluggable JS architecture developed for SilverStripe 4. The patterns it introduces, such as component registration, transformation and dependency injection between multiple JS bundles, could benefit projects in contexts outside of the SilverStripe admin interface - for example, the upcoming `silverstripe/mfa` module (which will contain a front-end React app that needs the ability to use React components from Authenticator plugin modules with their own independently compiled JS bundles).

Thankfully, Injector.js is already quite close to a drop-in state! It has no dependencies on other parts of the codebase, although we'll want to extract the Redux (`reducer`), Redux Form (`form`) & GraphQL (`query`) containers from it, since they're a lot less useful to other projects.

[Here's a PoC!](https://github.com/creative-commoners/outjector) This repo has an independent build of Injector.js, along with a 'primary' JS bundle (which acts as the 'host' app, like admin), and a 'secondary' bundle, which applies component registrations and transformations to 'primary' via Injector.js. See the README for details. This PoC also explores use of the `reducer` container, but the initial module should focus on supporting the React (`component`) container.

#### Considerations:

- The PoC build of Injector.js uses `expose-loader` to generate a bundle that applies Injector to the global scope, in order for the same Injector instance to be shared between different JS bundles. We'd need to determine what makes sense for the admin module:
  - Pulling in the dist version
  - Importing the source and exposing it via admin's own bundle
- We should try to keep the version history for the source files when they're shifted to a new repo - this'll take [a bit of wrangling](https://help.github.com/en/articles/splitting-a-subfolder-out-into-a-new-repository), but provided we get them all into one folder it should be doable.

#### Rough path to a shippable module:

- [ ] Extract Injector with the `component` container to a new repo
- [ ] Enable other containers to be registered externally
- [ ] Write some basic documentation for configuration and usage in standalone projects
- [ ] Publish to NPM
- [ ] Integrate external module with next minor release of SilverStripe (4.4 at this stage)

**Comment by robbieaverill** [comment]
@unclecheese @ScopeyNZ and I just had a chat about this in person and came up with some further thoughts:

* In admin at the moment, Injector's `loadComponent()` provides ApolloClient and Redux HOCs. Since these are primarily concerns of the admin/CMS area, these should remain as part of the silverstripe/admin module. They are not included in the current state of the react-injector library.
* In an imported react-injector world, you'd need to then import `provideApollo()` and `provideRedux()` for example from admin, then apply the HOCs explicitly when you export your component, rather than relying on Injector to do it when it loads your component.
* As above, for backwards compatibility we'd like to ensure that `loadComponent()` from admin still does this, while we deprecate it and suggest the new approach of using the less coupled HOCs.
* Injector would configure a global place to store its state, e.g. `window.ss.injector.default`, and modules could use a customised Injector store via something like `Injector.configure({ store: mycustomstore })`. In this case, any react-injector calls from that module would be resolved from `mycustomstore` rather than the `default` location.

We'd be doing this so that we can aim for a variety of packages being able to use react-injector _without_ admin dependency, and still share the same state. Currently this is achieved because admin exposes its injector and everything uses that, but in a "proper" `import { loadComponent } from 'react-injector';` world, the steps above would help us to achieve a common state store location for injector across modules.

**Comment by robbieaverill** [comment]
Given an example like this:

```jsx
class MyComponent extends Component {
  render() {
    const subComponent = loadComponent('MySubComponent');
    return <div><SubComponent /></div>;
  }
}

// ...

class MySubComponent extends Component {}
export default connect(mapStateToProps, mapDispatchToProps)(MySubComponent);
```

When running this in SilverStripe 4 with the silverstripe/admin provided Injector, the Redux store would be provided by `loadComponent()`. These modules would need to be changed to look like this:

```jsx
class MyComponent extends Component {
  render() {
    const subComponent = loadComponent('MySubComponent');
    return <div><SubComponent /></div>;
  }
}

// ...

class MySubComponent extends Component {}
export default compose(
  provideRedux,
  connect(mapStateToProps, mapDispatchToProps)
)(MySubComponent);
```

Does that sum up our conversation @ScopeyNZ?

**Comment by ScopeyNZ** [comment]
> how would a frontend-only module access the redux store without depending on silverstripe/admin?

Is "frontend" talking about JavaScript in the admin interface? If so, I think it's reasonable that you depend on admin, which provides redux. If it means "not part of admin" then I don't think you should be using the store from admin.

The only comment I'd make is that we already prefer using the `inject` HOC over `loadComponent` in React code - or using the context directly in the case of `FormBuilder`. `loadComponent` is really just used with entwine.

**Comment by robbieaverill** [comment]
Yeah sorry, I meant frontend as in "not part of admin". One of the reasons we'd be doing this is to support the case where three separate frontend (not part of admin) modules want to work in harmony with a react-injector that has a common store. Assuming we have redux here, and the `provideRedux` HOC is part of silverstripe/admin, we still have admin coupling to achieve that - right?
